Ensure NCCL error handling code is disabled for NCCL versions < 2.4

pritamdamania87 opened the pull request:
Summary:
ncclCommAbort() and ncclGetAsyncError() were two APIs added in NCCL
2.4 to detect errors in NCCL communicators. These were used as part of
ProcesGroupNCCL and we also enforced that only NCCL versions 2.4+ were
supported. Although, there is still legitimate use for older NCCL versions and
hence we should still support those.

For that purpose, in this change I've ensured we disable NCCL error checking
for versions < 2.4.

Test Plan:
1) Test with 2.4.8
2) Test with 2.2.13
3) unit tests.

Differential Revision: D17178988

pietern commented on test/common_distributed.py:
```diff
@@ -64,6 +64,11 @@ def requires_gloo():
+def requires_nccl_version(version):
+    return unittest.skipUnless(
+        torch.cuda.nccl.version() >= version,
+        "Requires nccl version greater than or equal to: {}, found: {}".format(version, torch.cuda.nccl.version()),
```
Capitalization of NCCL.

pietern commented on test/test_c10d.py:
```diff
@@ -2967,6 +2967,8 @@ def _test_broadcast_coalesced(self, process_group, device):
         pg.allreduce(torch.rand(10).cuda(self.rank))
 
+    # Need 2.4+ for error checking.
+    @requires_nccl_version(2400)
```
This check should be flipped with `requires_nccl`.

Could you add the comment to the skip check as argument, so it shows up in the skip description?

pietern reviewed (commented):
How does this impact the watchdog thread? Will it effectively be a big nop now?

LGTM otherwise.

mrshenli commented on test/test_c10d.py:
```diff
@@ -2967,6 +2967,8 @@ def _test_broadcast_coalesced(self, process_group, device):
 
+    # Need 2.4+ for error checking.
+    @requires_nccl_version(2400)
     @requires_nccl()
```
`requires_nccl` needs to be before `requires_nccl_version`? Or we can `c10d.is_nccl_available()` to `requires_nccl_version` as well.

pritamdamania87 commented:
> How does this impact the watchdog thread? Will it effectively be a big nop now?

Ensured we don't start this thread when error checking is disabled.

mrshenli reviewed (approved):
LGTM! Thanks! Let's make sure all CI pass before landing.

mrshenli commented on torch/lib/c10d/test/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -71,8 +72,8 @@ class ProcessGroupNCCLSimulateErrors : public c10d::ProcessGroupNCCL {
-    // Skip test if no cuda devices found.
-    return cudaNumDevices() == 0;
+    // Skip test if no cuda devices found or NCCL version < 2.4+.
+    return cudaNumDevices() == 0 || torch::cuda::nccl::version() < 2400;
```
This caused build failure:

```
Sep 05 20:06:57 CMakeFiles/ProcessGroupNCCLErrorsTest.dir/ProcessGroupNCCLErrorsTest.cpp.o: In function `ProcessGroupNCCLErrorsTest_testNCCLErrorsNonBlocking_Test::TestBody()':
Sep 05 20:06:57 ProcessGroupNCCLErrorsTest.cpp:(.text+0x4f1): undefined reference to `torch::cuda::nccl::version()'
```

pietern replied:
These files don't link against all of libtorch. It's better to use the macros here.

pritamdamania committed:
Ensure NCCL error handling code is disabled for NCCL versions < 2.4 (#25634)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25634

ncclCommAbort() and ncclGetAsyncError() were two APIs added in NCCL
2.4 to detect errors in NCCL communicators. These were used as part of
ProcesGroupNCCL and we also enforced that only NCCL versions 2.4+ were
supported. Although, there is still legitimate use for older NCCL versions and
hence we should still support those.

For that purpose, in this change I've ensured we disable NCCL error checking
for versions < 2.4.

Test Plan:
1) Test with 2.4.8
2) Test with 2.2.13
3) unit tests.

Differential Revision: D17178988

fbshipit-source-id: 757d4cf7c85dbfab804341b8b96fb480c4dda00b

jithunnair-amd commented:
@pritamdamania87 Will this PR be merged soon?

pritamdamania87 commented:
@jithunnair-amd Apologize for the delay on this, I was tied up with a few other PRs. I spent some time this week trying to fix the build issue but haven't made much progress on that front. Will give it another shot soon when I have some free cycles.

pritamdamania87 commented:
Abandoning in favor of https://github.com/pytorch/pytorch/pull/27124 (which was exported using ghexport).
